File: 20-actors/sanae/methods/test_labor_liberation.py
"""Tests for the LPS ranking method (ADR-2606032100)."""

from labor_liberation import (
    SEED_GAPS,
    SectorGap,
    displacement_cohort_size,
    freed_labor_hours,
    lps,
    ranked_seed,
)


def test_n1_excluded_sector_scores_zero():
    mining = next(g for g in SEED_GAPS if "MINING" in g.name)
    assert lps(mining) == 0.0  # charter_fit=0 zeroes the score (N1 gate proof)


def test_covered_sector_scores_below_uncovered_peer():
    # construction (coverage_gap 0.5, tatekata exists) vs garment (coverage_gap 1.0)
    constr = next(g for g in SEED_GAPS if "tatekata" in g.name)
    garment = next(g for g in SEED_GAPS if "hataori" in g.name)
    assert lps(garment) > lps(constr)


def test_sanae_ranks_first():
    assert "sanae" in ranked_seed()[0][0]  # largest headcount×misery, near-zero coverage


def test_wave_actors_in_top_band():
    # The wave OPENS sanae/hataori/kiyome (highest headcount×misery, zero coverage).
    # Note: raw automatability-weighting lifts kuramori (warehouse) into the top band too —
    # it is the #4/#5 roadmap item precisely because it is the most automatable of the
    # un-covered toil; the wave still prioritizes the zero-coverage sweatshop/cleaning first.
    top5 = {name for name, _ in ranked_seed()[:5]}
    assert any("sanae" in n for n in top5)
    assert any("hataori" in n for n in top5)
    assert any("kiyome" in n for n in top5)


def test_freed_hours_scales_linearly():
    assert freed_labor_hours(1000, 2000, 0.5) == 1_000_000


def test_cohort_size_rounds():
    assert displacement_cohort_size(1.0e6, 0.25) == 250_000


def test_higher_misery_raises_score_all_else_equal():
    base = SectorGap("x", "", "", "", 1e7, 2.0, 0.5, 1.0, 1.0)
    worse = SectorGap("y", "", "", "", 1e7, 3.0, 0.5, 1.0, 1.0)
    assert lps(worse) > lps(base)


if __name__ == "__main__":
    import sys
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    failed = 0
    for fn in fns:
        try:
            fn(); print(f"PASS {fn.__name__}")
        except AssertionError as e:
            failed += 1; print(f"FAIL {fn.__name__}: {e}")
    print(f"\n{len(fns)-failed}/{len(fns)} passed")
    sys.exit(1 if failed else 0)

File: 20-actors/sanae/methods/test_labor_liberation.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_labor_liberation.py.
(ns sanae.methods.test-labor-liberation
  "Tests for the LPS ranking method (ADR-2606032100).

  Run:  bb --classpath 20-actors 20-actors/sanae/methods/test_labor_liberation.clj"
  (:require [sanae.methods.labor-liberation :as ll]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(defn- by-name [frag] (first (filter #(str/includes? (:name %) frag) ll/seed-gaps)))

(deftest n1-excluded-sector-scores-zero
  (is (= 0.0 (ll/lps (by-name "MINING")))))   ; charter-fit=0 zeroes the score (N1 gate proof)

(deftest covered-sector-scores-below-uncovered-peer
  ;; construction (coverage-gap 0.5, tatekata exists) vs garment (coverage-gap 1.0)
  (is (> (ll/lps (by-name "hataori")) (ll/lps (by-name "tatekata")))))

(deftest sanae-ranks-first
  (is (str/includes? (first (first (ll/ranked-seed))) "sanae")))

(deftest wave-actors-in-top-band
  (let [top5 (set (map first (take 5 (ll/ranked-seed))))]
    (is (some #(str/includes? % "sanae") top5))
    (is (some #(str/includes? % "hataori") top5))
    (is (some #(str/includes? % "kiyome") top5))))

(deftest freed-hours-scales-linearly
  (is (= 1000000.0 (ll/freed-labor-hours 1000 2000 0.5))))

(deftest cohort-size-rounds
  (is (= 250000 (ll/displacement-cohort-size 1.0e6 0.25))))

(deftest higher-misery-raises-score-all-else-equal
  (let [base  (ll/sector-gap "x" "" "" "" 1e7 2.0 0.5 1.0 1.0)
        worse (ll/sector-gap "y" "" "" "" 1e7 3.0 0.5 1.0 1.0)]
    (is (> (ll/lps worse) (ll/lps base)))))

(deftest fishing-sector-present
  ;; cross-link to umisachi 海幸 — the marine-bounty actor that bodies this fishing gap
  (is (some? (by-name "ama (fishing")))
  (is (pos? (ll/lps (by-name "ama (fishing")))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'sanae.methods.test-labor-liberation)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
